File: home/sensor.py
import random
from home.logger_config import logger


# todo: we do not need to specify 'room' in the logger msg, if we already have the 'id'
class Sensor:
    sensor_count = {}  # Dictionary to keep track of sensor count for each room and sensor type

    def __init__(self, sensor_type, room_name):
        if room_name not in Sensor.sensor_count:
            Sensor.sensor_count[room_name] = {}  # Initialize sensor count for the room if not already present
        if sensor_type not in Sensor.sensor_count[room_name]:
            Sensor.sensor_count[room_name][
                sensor_type] = 0  # Initialize sensor count for the sensor type if not already present

        Sensor.sensor_count[room_name][sensor_type] += 1  # Increment sensor count for the sensor type in the room

        self.id = f"/Sensor/{sensor_type}/{room_name}/{Sensor.sensor_count[room_name][sensor_type]}"
        self.sensor_type = sensor_type
        self.room_name = room_name
        self.status = "off"

    def turn_on(self):
        self.status = "on"
        print(f"'{self.id}' is now {self.status.upper()}.")
        logger.info(format("\'" + self.id + "\' is turned " + self.status.upper()))

    def turn_off(self):
        self.status = "off"
        print(f"{self.id} is now OFF.")
        logger.info(format("\'" + self.id + "\' is turned " + self.status.upper()))

    def get_reading(self):
        if self.status == "off":
            print(f"Cannot Get Sensor Reading, {self.id} is Currently OFF. ")
            logger.warning(format("\'" + self.id + "\' is currently OFF. Cannot get reading."))
            return None
        elif self.status == "on":
            reading = self._generate_reading()
            print(f"{self.id} reading is: {reading}")
            logger.info(format("\'" + self.id + "\' Reading is: " + str(reading)))
            return reading
        else:
            logger.error(format("\'" + self.id + "\' is in error status."))
            print(f"Status Error with status: {self.status}")

    def get_status(self):
        print(f"{self.id} current status is: {self.status}")
        logger.info(format("\'" + self.id + "\' CURRENT STATUS is: " + self.status.upper()))
        return self.status

    def _generate_reading(self):
        raise NotImplementedError("Subclasses must implement _generate_reading method.")


class IndoorTemperatureSensor(Sensor):
    def __init__(self, room_name):
        super().__init__("IndoorTemperature", room_name)

    def _generate_reading(self):
        return round(random.uniform(30, 40), 2)


class OutdoorTemperatureSensor(Sensor):
    def __init__(self, room_name):
        super().__init__("OutdoorTemperature", room_name)

    def _generate_reading(self):
        return round(random.uniform(20, 25), 2)


class HumiditySensor(Sensor):
    def __init__(self, room_name):
        super().__init__("Humidity", room_name)

    def _generate_reading(self):
        return round(random.uniform(12, 30), 2)


class LightIntensiveSensor(Sensor):
    def __init__(self, room_name):
        super().__init__("LightIntensive", room_name)

    def _generate_reading(self):
        return round(random.uniform(900, 1200), 2)


class SmokeSensor(Sensor):
    def __init__(self, room_name):
        super().__init__("Smoke", room_name)

    def _generate_reading(self):
        return round(random.uniform(0, 100), 2)


if __name__ == "__main__":
    # pass

    uv_sensor = LightIntensiveSensor("test")
    uv_sensor.turn_on()
    uv_sensor.get_status()
    print(uv_sensor.get_reading())
    uv_sensor.turn_off()
    print(uv_sensor.get_reading())

    outdoor_temp_sensor = OutdoorTemperatureSensor("test")
    outdoor_temp_sensor.turn_on()
    print(outdoor_temp_sensor.get_reading())

File: experiments_js/home/sensor.js
const { logger } = require('./logger_config');

class Sensor {
    static sensorCount = {}; // Track sensor count for each room and sensor type

    constructor(sensorType, roomName) {
        if (!Sensor.sensorCount[roomName]) {
            Sensor.sensorCount[roomName] = {}; // Initialize room if not already present
        }
        if (!Sensor.sensorCount[roomName][sensorType]) {
            Sensor.sensorCount[roomName][sensorType] = 0; // Initialize sensor type count
        }

        Sensor.sensorCount[roomName][sensorType]++; // Increment sensor count

        this.id = `/Sensor/${sensorType}/${roomName}/${Sensor.sensorCount[roomName][sensorType]}`;
        this.sensorType = sensorType;
        this.roomName = roomName;
        this.status = "off";
    }

    turnOn() {
        this.status = "on";
        console.log(`'${this.id}' is now ${this.status.toUpperCase()}.`);
        logger.info(`${this.id} is turned ${this.status.toUpperCase()}.`);
    }

    turnOff() {
        this.status = "off";
        console.log(`${this.id} is now OFF.`);
        logger.info(`${this.id} is turned ${this.status.toUpperCase()}.`);
    }

    getStatus() {
        console.log(`${this.id} current status is: ${this.status}`);
        logger.info(`${this.id} CURRENT STATUS is: ${this.status.toUpperCase()}`);
        return this.status;
    }

    getReading() {
        if (this.status === "off") {
            console.log(`Cannot Get Sensor Reading, ${this.id} is Currently OFF.`);
            logger.warning(`${this.id} is currently OFF. Cannot get reading.`);
            return null;
        } else if (this.status === "on") {
            const reading = this._generateReading();
            console.log(`${this.id} reading is: ${reading}`);
            logger.info(`${this.id} Reading is: ${reading}`);
            return reading;
        } else {
            console.log(`Status Error with status: ${this.status}`);
            logger.error(`${this.id} is in error status`);
            return null;
        }

    }

    _generateReading() {
        throw new Error("Subclasses must implement _generateReading.");
    }
}

class IndoorTemperatureSensor extends Sensor {
    constructor(roomName) {
        super("IndoorTemperature", roomName);
    }

    _generateReading() {
        return parseFloat((Math.random() * (40 - 30) + 30).toFixed(2)); // Random temp between 30-40°C
    }
}

class OutdoorTemperatureSensor extends Sensor {
    constructor(roomName) {
        super("OutdoorTemperature", roomName);
    }

    _generateReading() {
        return parseFloat((Math.random() * (25 - 20) + 20).toFixed(2)); // Random temp between 20 and 25°C
    }
}

class HumiditySensor extends Sensor {
    constructor(roomName) {
        super("Humidity", roomName);
    }

    _generateReading() {
        return parseFloat((Math.random() * (30 - 12) + 12).toFixed(2)); // Random humidity between 12-30%
    }
}

class LightIntensiveSensor extends Sensor {
    constructor(roomName) {
        super("LightIntensive", roomName);
    }

    _generateReading() {
        return parseFloat((Math.random() * (1200 - 900) + 900).toFixed(2)); // Random light intensity between 900-1200 lux
    }
}

class SmokeSensor extends Sensor {
    constructor(roomName) {
        super("Smoke", roomName);
    }

    _generateReading() {
        return parseFloat((Math.random() * 100).toFixed(2)); // Random smoke level between 0-100
    }
}

module.exports = { IndoorTemperatureSensor, OutdoorTemperatureSensor, HumiditySensor, LightIntensiveSensor, SmokeSensor };